File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.java
class Solution {
  public int[] topKFrequent(int[] nums, int k) {
    Map<Integer, Integer> frequencies = new HashMap<>();
    for (int num : nums) {
      frequencies.put(num, frequencies.getOrDefault(num, 0) + 1);
    }
    PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
    for (Map.Entry<Integer, Integer> entry : frequencies.entrySet()) {
      if (pq.size() == k) {
        pq.offer(new int[] { entry.getValue(), entry.getKey() });
        pq.poll();
      } else {
        pq.offer(new int[] { entry.getValue(), entry.getKey() });
      }
    }
    int[] ans = new int[k];
    int i = 0;
    for (int[] pair : pq) {
      ans[i] = pair[1];
      i++;
    }
    return ans;
  }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixes = []
        suffixes = []
        ans = []
        prefixes.append(1)
        tmp = nums[0]
        for num in nums[1::]:
            prefixes.append(tmp)
            tmp = tmp*num
        suffixes.append(1)
        tmp = nums[-1]
        for num in nums[-2::-1]:
            suffixes.append(tmp)
            tmp = tmp*num
        suffixes = suffixes[::-1]
        print(prefixes)
        print(suffixes)
        for i in range(len(prefixes)):
            ans.append(prefixes[i]*suffixes[i])
        return ans

